### PR TITLE
Remove finalizer code that was never reachable

### PR DESCRIPTION
**Pull request by alex**

### CODE REVIEW

**Comment by kngwyu** [comment]
I'd rather complete the new `tp_dealloc` in a PR to ensure it works.

**Comment by alex** [comment]
What's the new `tp_dealloc`?

**Comment by kngwyu** [comment]
I meant that we need to update this dealloc function not to use `tp_free`, so the related changes should packed in a PR.

**Comment by alex** [comment]
I haven't figured out how to do that yet :-) So far all the tests pass if I simply always call `tp_free_fallback` instead of `Py_TYPE(obj).tp_free`, but I'm not sure it's right.

**Comment by alex** [comment]
@kngwyu I have just discovered `PyType_GetSlot`... which almost works, except in the `tp_new` case, because you can only use `PyType_GetSlot` with heap types. What do you think about disabling inheriting from builtin types with limited API?

**Comment by alex** [comment]
Looks like PyPy support for `PyType_GetSlot` is new in 7.3.2, but I can't figure out which version of PyPy is used in CI.

**Review by kngwyu — CHANGES_REQUESTED** [review]
Thanks, this approach looks correct.

PyPy 7.3.2 is the latest version and I think supporting only the latest PyPy is not good.
So I think it's good to allow accessing `tp_free` or so only for PyPy, by defining some helper functions, though I actually don't know how PyPy treats LIMITED_API.

File: src/pyclass.rs [hunk 1]
```diff
@@ -46,13 +55,11 @@ pub trait PyClassAlloc: PyTypeInfo + Sized {
...
-        }
 
-        match (*ffi::Py_TYPE(obj.as_ptr())).tp_free {
+        match mem::transmute::<_, Option<ffi::freefunc>>(ffi::PyType_GetSlot(
+            ffi::Py_TYPE(obj.as_ptr()),
+            ffi::Py_tp_free,
+        )) {
             Some(free) => free(obj.as_ptr() as *mut c_void),
             None => tp_free_fallback(obj.as_ptr()),
         }
```
[the remark this answers is not in the document]
Comment (alex) [reply src/pyclass.rs:65]: Was this always a bug, or a new one?

[the remark this answers is not in the document]
Comment (kngwyu) [reply src/pyclass.rs:80]: New one introduced by the use of `PyTypeObject_FromSpec`.
Currently (on master), type object is allocated by Rust and leaked.

File: src/pyclass.rs [hunk 2]
```diff
@@ -46,14 +75,13 @@ pub trait PyClassAlloc: PyTypeInfo + Sized {
...
-            return;
-        }
 
-        match (*ffi::Py_TYPE(obj.as_ptr())).tp_free {
-            Some(free) => free(obj.as_ptr() as *mut c_void),
+        match get_type_free(ffi::Py_TYPE(obj.as_ptr())) {
+            Some(free) => {
+                let ty = ffi::Py_TYPE(obj.as_ptr());
+                free(obj.as_ptr() as *mut c_void);
+                ffi::Py_DECREF(ty as *mut ffi::PyObject);
```
Comment (kngwyu) [inline src/pyclass.rs:83]: I'm sorry for my confusing comment, but this `DECREF` should be always called.

Comment (alex) [reply src/pyclass.rs:83]: `tp_free_fallback` already does it.

**Comment by alex** [comment]
Ok, tests are green on PyPy now.

**Review by kngwyu — APPROVED** [review]

**Comment by kngwyu** [comment]
Thank you!

**Pull request merged**